registry/logging.go

```diff
@@ -0,0 +1,118 @@
+package registry
+
+import (
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	jsoniter "github.com/json-iterator/go"
+)
+
+var json = jsoniter.ConfigCompatibleWithStandardLibrary
+
+// LogEntry represents a log entry in JSON format.
+type JSONLogEntry struct {
+	Timestamp string `json:"timestamp"`
+	Method    string `json:"method"`
+	Path      string `json:"path"`
+	Status    int    `json:"status"`
+	Size      int    `json:"size"`
+	Referer   string `json:"referer"`
+	UserAgent string `json:"user_agent"`
+}
+
+// jsonLogFormatterParams holds the parameters required for JSON logging.
+type JSONLogFormatterParams struct {
+	Request   *http.Request
+	URL       *url.URL
+	Timestamp time.Time
+	Status    int
+	Size      int
+}
+
+type responseLogger struct {
+	w      http.ResponseWriter
+	status int
+	size   int
+}
+
+func (l *responseLogger) Write(p []byte) (int, error) {
+	size, err := l.w.Write(p)
+	l.size += size
+	return size, err
+}
+
+func (l *responseLogger) WriteHeader(status int) {
+	l.w.WriteHeader(status)
+	l.status = status
+}
+
+func (l *responseLogger) Status() int {
+	if l.status == 0 {
+		return http.StatusOK
+	}
+	return l.status
+}
+
+func (l *responseLogger) Size() int {
+	return l.size
+}
+
+func makeLogger(w http.ResponseWriter) (*responseLogger, http.ResponseWriter) {
+	logger := &responseLogger{w: w, status: http.StatusOK}
+	return logger, w
+}
+
+// writeJSONCombinedLog writes a log entry for req to w in JSON format similar to Combined Log Format.
+func writeJSONCombinedLog(writer io.Writer, params JSONLogFormatterParams) {
+	entry := JSONLogEntry{
+		Timestamp: params.Timestamp.Format("02/Jan/2006:15:04:05 -0700"),
+		Method:    params.Request.Method,
+		Path:      params.URL.Path,
+		Status:    params.Status,
+		Size:      params.Size,
+		Referer:   params.Request.Referer(),
+		UserAgent: params.Request.UserAgent(),
+	}
+
+	// Use jsoniter to marshal the struct
+	jsonEntry, err := json.Marshal(entry)
+	if err != nil {
+		// Handle error
+		return
+	}
+
+	jsonEntry = append(jsonEntry, '\n')
+	_, _ = writer.Write(jsonEntry)
+}
+
+// JSONLoggingHandler returns a http.Handler that wraps h and logs requests in JSON
+// format similar to Combined Log Format.
+func JSONLoggingHandler(out io.Writer, h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
```

Comment:
If we decide to use an encoder; we can probably instance it here ("once"), and pass it to `writeJSONCombinedLog` when handling requests.